hoodie-client/src/main/java/com/uber/hoodie/table/HoodieCopyOnWriteTable.java

```diff
@@ -479,4 +482,110 @@ protected HoodieUpdateHandle getUpdateHandle(String commitTime, String fileLoc,
+                return partitionCleanStatsMap.get(partitionPath);
+            } else {
+                return HoodieCleanStat.newBuilder().withPolicy(config.getCleanerPolicy())
+                    .withPartitionPath(partitionPath)
+                    .withDeletePathPattern(new ArrayList<>())
+                    .withEarliestCommitRetained(cleaner.getEarliestCommitToRetain())
+                    .withDeletedFileResults(new HashMap<>()).build();
+            }
+        }).collect(Collectors.toList());
+    }
+
+    private PairFlatMapFunction<Iterator<Tuple2<String, String>>, String, HoodieCleanStat> deleteFilesFunc(
+        HoodieTable table, HoodieWriteConfig config) {
+        return (PairFlatMapFunction<Iterator<Tuple2<String, String>>, String, HoodieCleanStat>) iter -> {
+            HoodieCleaner cleaner = new HoodieCleaner(table, config);
+            Map<String, HoodieCleanStat> partitionCleanStatMap = new HashMap<>();
+
+            while (iter.hasNext()) {
+                Tuple2<String, String> partitionDelFileTuple = iter.next();
+                String partitionPath = partitionDelFileTuple._1();
+                String deletePathStr = partitionDelFileTuple._2();
+                Boolean deletedFileResult = deleteFileAndGetResult(deletePathStr);
+                if (!partitionCleanStatMap.containsKey(partitionPath)) {
+                    partitionCleanStatMap.put(partitionPath,
+                        HoodieCleanStat.newBuilder()
+                            .withPolicy(config.getCleanerPolicy())
+                            .withPartitionPath(partitionPath)
+                            .withEarliestCommitRetained(
+                                cleaner.getEarliestCommitToRetain()).build());
+                }
+                partitionCleanStatMap.get(partitionPath)
+                    .addDeletePathPattern(deletePathStr)
+                    .addDeletedFileResult(deletePathStr, deletedFileResult);
+            }
+
+            return partitionCleanStatMap.entrySet().stream()
+                .map(e -> new Tuple2<>(e.getKey(), e.getValue()))
+                .collect(Collectors.toList()).iterator();
+        };
+    }
+
+    private static PairFlatMapFunction<String, String, String> getFilesToDeleteFunc(
+        HoodieTable table, HoodieWriteConfig config) {
+        return (PairFlatMapFunction<String, String, String>) partitionPathToClean -> {
+            HoodieCleaner cleaner = new HoodieCleaner(table, config);
+            return cleaner.getDeletePaths(partitionPathToClean).stream()
+                .map(deleteFile -> new Tuple2<>(partitionPathToClean, deleteFile.toString()))
+                .iterator();
+        };
+    }
+
+    private Boolean deleteFileAndGetResult(String deletePathStr) throws IOException {
+        Path deletePath = new Path(deletePathStr);
+        logger.info("Working on delete path :" + deletePath);
+        boolean deleteResult = getFs().delete(deletePath, false);
```

Comment:
If HoodieTable does the actual clean, then HoodieCleaner is confusing to me. May be rename it to HoodieCleanHelper